include/soci/row.h

```diff
@@ -126,9 +134,77 @@ class SOCI_DECL row
 
     std::size_t find_column(std::string const& name) const;
 
+    std::pair<std::string*, indicator*> alloc_data_holder_string();
+    std::pair<std::tm*, indicator*> alloc_data_holder_tm();
+    std::pair<double*, indicator*> alloc_data_holder_double();
+    std::pair<int*, indicator*> alloc_data_holder_int();
+    std::pair<long long*, indicator*> alloc_data_holder_llong();
+    std::pair<unsigned long long*, indicator*> alloc_data_holder_ullong();
+
+    data_type get_data_holder_type(std::size_t pos) const SOCI_NOEXCEPT;
+    double get_data_holder_double(std::size_t pos) const SOCI_NOEXCEPT;
+    int get_data_holder_int(std::size_t pos) const SOCI_NOEXCEPT;
+    long long get_data_holder_llong(std::size_t pos) const SOCI_NOEXCEPT;
+    unsigned long long get_data_holder_ullong(std::size_t pos) const SOCI_NOEXCEPT;
+
+    template<typename Dst>
+    struct numeric_cast_t
+    {
+        template<typename Src>
+        typename soci::enable_if<soci::is_same<Dst, Src>::value, Dst>::type
+        operator()(Src src) const
+        {
+            return src;
+        }
+#ifdef SOCI_HAVE_BOOST
+        template<typename Src>
+        typename soci::enable_if<!soci::is_same<Dst, Src>::value, Dst>::type
+        operator()(Src src) const
+        {
+            return boost::numeric_cast<Dst>(src);
+        }
+#else
+        template<typename Src>
+        typename soci::enable_if<!soci::is_same<Dst, Src>::value, Dst>::type
+        operator()(Src) const
+        {
+            throw std::bad_cast();
+        }
+#endif
+    };
+
+    void get_(std::size_t pos, std::string &baseVal) const;
+    void get_(std::size_t pos, std::tm &baseVal) const;
+    template <typename T>
+    typename soci::enable_if<std::numeric_limits<T>::is_specialized, void>::type
+    get_(std::size_t pos, T &baseVal) const
+    {
+        return get_number(pos, baseVal, numeric_cast_t<T>());
+    }
+    template <typename T, typename NumericCast>
+    void get_number(std::size_t pos, T &baseVal, const NumericCast &cast) const
```

Comment:
It looks like this function is only useful if we're using Boost, as otherwise it never performs any non-trivial casts? If this is correct, I don't know if it's a good idea and I'd rather never cast between the different numbers rather than only doing it when Boost is available but not without it.

Does it already work like this in the current version of SOCI? If so, I had no idea about it...

Reply:
I have tried to implement numeric_cast<> use only c++ standard library. But the compiler warnings are not easy to fix. Although  SOCI::into doesn't suffer the number conversion problem, I should use SOCI::into instead, but ORM depends on SOCI::row and some functions in different databases generate different data types. So soci::row with auto numeric conversion is very implement to me. I also know the difference between WITH_BOOST 'ON' and WITH_BOOST 'OFF'  is ugly, but I have not found a better way, so I use number_cast just for BOOST 'ON' and leave numeric_cast<> for BOOST 'OFF'  the future work.<br/>

Do you have any good ideas for me? Because number_cast<> is only in row.h, So I think remove number_cast from row.h and patch this only when I use it in my project may be a compromise way.